### ISSUE TITLE
Mac Retina icon resize problem

### ISSUE THREAD

**Issue by dyorgio** [dorkbox/SystemTray#51]
EDIT: Clarifing the problem: Icons are blurry on Mac retina using SystemTray, using AWT they are crisp.

Not sure what is causing this, ImageUtils.TRAY_SIZE= 32;

when use SystemTray lib:
<img width="67" alt="captura de tela 2017-03-01 as 18 43 24" src="https://cloud.githubusercontent.com/assets/1438251/23482787/17bd6566-feaf-11e6-83a5-6fa7322e8161.png">

when use AWT directly:
<img width="58" alt="captura de tela 2017-03-01 as 18 44 04" src="https://cloud.githubusercontent.com/assets/1438251/23482797/1c5bc8ec-feaf-11e6-989b-cca5416e2b2f.png">

when use SystemTray lib with AUTO_TRAY_SIZE = false:
<img width="60" alt="captura de tela 2017-03-01 as 18 50 01" src="https://cloud.githubusercontent.com/assets/1438251/23483024/f80eca1a-feaf-11e6-8a8c-fcef4abbad5c.png">

Original file:
![icon72-mactray 2x](https://cloud.githubusercontent.com/assets/1438251/23482813/29e418c0-feaf-11e6-9d20-dbe0e75bc1e6.png)

**Comment by dyorgio** [comment]
The problem disappears if I set manually default tray size var:

```java
dorkbox.systemTray.SystemTray.DEFAULT_TRAY_SIZE = 64;
```

The AUTO_TRAY_SIZE value doesn't change final result.

**Comment by dorkbox** [comment]
I'm not sure if I understand what the problem is.

if `AUTO_TRAY_SIZE = false`, then there is no automatic scaling for the tray icon, which means you must provide the correct default size via `DEFAULT_TRAY_SIZE`.

You can see what is going on here: https://github.com/dorkbox/SystemTray/blob/master/src/dorkbox/systemTray/util/ImageUtils.java#L233

**Comment by dyorgio** [comment]
Hi @dorkbox, the problem is that auto tray size is not working correctly on macOS with retina display.
ImageUtils code needs to finish auto size logic with TRAY_SIZE = 64, currently it is finish with 32.

**Comment by dyorgio** [comment]
@dorkbox I believe that the problem is using windows sizes in macOS...

windows = 16x16, 32x32 (hdpi)
macOS = 22x22, 44x44 (retina)

it makes any sense?

**Comment by dorkbox** [comment]
I think I understand.

It's not so much that it's using windows sizes (if you look at the code, it's completely separate logic), but rather that java/awt isn't reporting the correct size.

I'm still confused about your first post though...
> when use SystemTray lib:
<img width="67" alt="captura de tela 2017-03-01 as 18 43 24" src="https://cloud.githubusercontent.com/assets/1438251/23482787/17bd6566-feaf-11e6-83a5-6fa7322e8161.png">
(this image looks like it is the correct size to me)

**Comment by dyorgio** [comment]
The problem is not the size :), the problem is in resize, the first image is smooth compared with pure AWT trayicon version.
size 44 is also make my icon blurry... :(, Can I disable resize completely?

**Comment by dorkbox** [comment]
so if I understand correctly...
`DEFAULT_TRAY_SIZE = 64;` + `AUTO_TRAY_SIZE = false;` = perfect icon in the tray?

**Comment by dorkbox** [comment]
Also, is your initial icon a power of 2, ie: 16, 32, or 64 -- scaling/resize doesn't work as well when it's a non-power of 2.

**Comment by dorkbox** [comment]
And the last question... 
> the first image is smooth

Isn't this what you want?

**Comment by dorkbox** [comment]
I think I understand the problem.. `AUTO_TRAY_SIZE = false` will use the size of the specified icon (instead of using the `DEFAULT_TRAY_SIZE` to determine it's size. 

This will be fixed in the next release.

**Comment by dyorgio** [comment]
Tks @dorkbox! I believe that it solves the problem :)

**Comment by dyorgio** [comment]
For future references, my current code:

```java
if (OSDetector.isMac()) {
    dorkbox.systemTray.SystemTray.AUTO_TRAY_SIZE = false;
    if (isHdpi()){
        dorkbox.systemTray.SystemTray.DEFAULT_TRAY_SIZE = 44;
    } else {
        dorkbox.systemTray.SystemTray.DEFAULT_TRAY_SIZE = 22;
    }
}
```

**Comment by dorkbox** [comment]
How do you determine isHdpi()?

**Comment by dyorgio** [comment]
Hi @dorkbox, my magic code (a join of solutions found on internet):

```java
private static Boolean cachedIsHdpi = null;

public static boolean isHdpi() {
    if (cachedIsHdpi != null) {
        return cachedIsHdpi;
    }

    boolean is2xScale = is2xScale();
    if (!is2xScale) {
        GraphicsDevice graphicsDevice = GraphicsEnvironment.getLocalGraphicsEnvironment().getDefaultScreenDevice();

        try {
            Field field = graphicsDevice.getClass().getDeclaredField("scale");
            if (field != null) {
                field.setAccessible(true);
                Object scale = field.get(graphicsDevice);
                if (scale instanceof Integer && ((Integer) scale) == 2) {
                    is2xScale = true;
                }
            }
        } catch (Exception e) {
        }
    }
    return cachedIsHdpi = is2xScale;
}

private static boolean is2xScale() {
    Object obj = Toolkit.getDefaultToolkit().getDesktopProperty("apple.awt.contentScaleFactor");
    if (obj != null && obj instanceof Float) {
        Float f = (Float) obj;
        int scale = f.intValue();
        return (scale == 2);
    } else {
        BufferedImage img = new BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = GraphicsEnvironment.getLocalGraphicsEnvironment().createGraphics(img);
        return g.getFontRenderContext().getTransform().equals(AffineTransform.getScaleInstance(2.0, 2.0));
    }
}
```

**Comment by dorkbox** [comment]
Cool. That looks to be a lot better detection than what I'm currently using. Do I have your permission to use this under Apache 2.0?

**Comment by dyorgio** [comment]
Of course :), open source rules!

**Comment by dorkbox** [comment]
MacOS fixed in 686e14401e50b50aa11afb457b45217d1fcf7a5d (warning, there was a lot going on when this was fixed....)
